discovery/dns/dns.go

```diff
@@ -169,39 +170,117 @@ func (d *Discovery) refresh(ctx context.Context, name string, ch chan<- []*confi
 	response := &dns.Msg{}
+	allResponsesValid := true
+
+	for _, lname := range conf.NameList(name) {
+		response, err = lookupName(lname, qtype, conf, logger)
+
+		if err != nil {
+			// We can't go home yet, because a later name
+			// may give us a valid, successful answer.  However
+			// we can no longer say "this name definitely doesn't
+			// exist", because we did not get that answer for
+			// at least one name.
+			allResponsesValid = false
+		} else if response.Rcode == dns.RcodeSuccess {
+			// Outcome 1: GOLD!
+			return response, nil
+		}
+	}
+
+	if allResponsesValid {
+		// Outcome 2: everyone says NXDOMAIN, that's good enough for me
+		return nil, nil
+	}
+	// Outcome 3: boned.
+	return nil, fmt.Errorf("could not resolve %s: one or more search-path-mangled names returned all errors", name)
+}
+
+// Lookup a specific name across all configured servers.  If a viable answer
+// is received from a server, then it is immediately returned, otherwise the
+// other servers in the config are tried, and if none of them return a
+// viable answer, an error is returned.
+//
+// A "viable answer" is one which indicates either:
+//
+// 1. "yes, I know that name, and here are its records of the requested type"
+//    (RCODE==SUCCESS, ANCOUNT > 0);
+// 2. "yes, I know that name, but it has no records of the requested type"
+//    (RCODE==SUCCESS, ANCOUNT==0); or
+// 3. "I know that name doesn't exist" (RCODE==NXDOMAIN).
+//
+// A non-viable answer is "anything else", which encompasses both various
+// system-level problems (like network timeouts) and also
+// valid-but-unexpected DNS responses (SERVFAIL, REFUSED, etc).
+//
+func lookupName(name string, qtype uint16, conf *dns.ClientConfig, logger log.Logger) (*dns.Msg, error) {
+	client := &dns.Client{}
 
 	for _, server := range conf.Servers {
 		servAddr := net.JoinHostPort(server, conf.Port)
-		for _, lname := range conf.NameList(name) {
-			response, err = lookup(lname, qtype, client, servAddr, false)
-			if err != nil {
-				logger.
-					With("server", server).
-					With("name", name).
-					With("reason", err).
-					Warn("DNS resolution failed.")
-				continue
-			}
-			if len(response.Answer) > 0 {
-				return response, nil
-			}
+		msg, err := askServerForName(name, qtype, client, servAddr, false)
+		if err != nil {
+			logger.
+				With("server", server).
+				With("name", name).
+				With("reason", err).
+				Warn("DNS resolution failed.")
+			continue
+		}
+
+		if msg.Rcode == dns.RcodeSuccess || msg.Rcode == dns.RcodeNameError {
+			// We have our answer.  Time to go home.
+			return msg, nil
 		}
 	}
-	return response, fmt.Errorf("could not resolve %s: no server responded", name)
+
+	return nil, fmt.Errorf("could not resolve %s: no servers returned a viable answer", name)
 }
 
-func lookup(lname string, queryType uint16, client *dns.Client, servAddr string, edns bool) (*dns.Msg, error) {
+// Make a request to a specific DNS server for a specific name (and qtype).
```

Comment:
Follow the go docstring style and start the sentence with the function name `askServerForName makes a request ...`.